### ISSUE TITLE
Remote signer

### ISSUE THREAD

**Issue by ligi** [walleth/walleth#149]
Let WALLETH act as a remote signer and use remote signers. These are use-cases:

 * Be a remote signer for a geth node (analog to https://github.com/holiman/qtsigner)
 * Browse web3 websites with your ("dirty") browser - but have your keys and signing interaction on a device/context with smaller attack-surface.
 * more sophisticated hardware wallet
 * ..

Open questions:
 - communication consumer<>signer (USB?,PSS?,..)

**Comment by holiman** [comment]
For "my" usecase, which is "Be a signer for a geth", I envision it like this: 

- walleth (or a different android app? Let's call it w2) bundles the go-implementation of the signer. The android application would use the in-proc go interface to 'hook' up the right methods (not based on intents). Just like the 'laptop'-signer natively has a command line UI, the 'mobile' signer would natively have an android UI. 
- w2 would open the following external interfaces, both considered untrusted:
  - port 80, to listen to signing requests. Useful when your laptop and phone are on the same network. 
  - intents, also to listen to signing requests. 

This would allow other applications to get transactions signed, as well as laptops. Further interfaces could be added, e.g. QR.

**Comment by ligi** [comment]
> - walleth (or a different android app? Let's call it w2)

I am creating a flavor for this. This basically means it can be a different android app (can build different apk, package name and app-name if needed) - Just noticed there will be a lot of shared elements (Account management(import/export/creation/..) and want to be DRY)
 
> bundles the go-implementation of the signer. The android application would use the in-proc go interface to 'hook' up the right methods (not based on intents). Just like the 'laptop'-signer natively has a command line UI, the 'mobile' signer would natively have an android UI. 

ACK - sounds good

> - w2 would open the following external interfaces, both considered untrusted:
>   - port 80, to listen to signing requests. Useful when your laptop and phone are on the same network. 

Not yet sure if this is really the way to go in the future - I see horrible UX down this path. But it can be a way to go for the first PoCs

>  - intents, also to listen to signing requests. 

YAY - this is where I want to go really - should also standardize this interface via ERC 

> This would allow other applications to get transactions signed, as well as laptops. Further interfaces could be added, e.g. QR. 

ACK - but I would do QR very last - the UX is really bad - but can also standardize and extend the offline transaction functionality

**Comment by holiman** [comment]
> YAY - this is where I want to go really - should also standardize this interface via ERC

Yes, but I think we need to iterate on both the signer<->UI interface, aswell as the external interface (which will impact the intent-interface) a few more rounds before we're at ERC stage .

**Comment by ligi** [comment]
> Yes, but I think we need to iterate on both the signer<->UI interface, aswell as the external interface (which will impact the intent-interface) a few more rounds before we're at ERC stage .

ACK - but let's not delay it to far away - IMHO an ERC is a good way to get valuable comments and feedback with potential use-cases - it can be a WIP-ERC with not much focus on formalization in the beginning - just to get external input early
